lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/operators/ScatterReceiver.java

```diff
@@ -0,0 +1,162 @@
+
+  private final int version;
+
+  @Inject
+  public ScatterReceiver(@Parameter(CommunicationGroupName.class) final String groupName,
+                         @Parameter(OperatorName.class) final String operName,
+                         @Parameter(TaskConfigurationOptions.Identifier.class) final String selfId,
+                         @Parameter(DataCodec.class) final Codec<T> dataCodec,
+                         @Parameter(DriverIdentifier.class) final String driverId,
+                         @Parameter(TaskVersion.class) final int version,
+                         final CommGroupNetworkHandler commGroupNetworkHandler,
+                         final NetworkService<GroupCommunicationMessage> netService,
+                         final CommunicationGroupServiceClient commGroupClient) {
+    super();
+    LOG.log(Level.FINEST, "{0} has CommGroupHandler-{1}", new Object[]{operName, commGroupNetworkHandler});
+    this.version = version;
+    this.groupName = Utils.getClass(groupName);
+    this.operName = Utils.getClass(operName);
+    this.dataCodec = dataCodec;
+    this.commGroupNetworkHandler = commGroupNetworkHandler;
+    this.netService = netService;
+    this.sender = new Sender(this.netService);
+    this.topology = new OperatorTopologyImpl(this.groupName, this.operName, selfId, driverId, sender, version);
+    this.selfId = selfId;
+    this.commGroupNetworkHandler.register(this.operName, this);
+    this.commGroupClient = commGroupClient;
+  }
+
+  @Override
+  public int getVersion() {
+    return version;
+  }
+
+  @Override
+  public void initialize() throws ParentDeadException {
+    topology.initialize();
+  }
+
+  @Override
+  public Class<? extends Name<String>> getOperName() {
+    return operName;
+  }
+
+  @Override
+  public Class<? extends Name<String>> getGroupName() {
+    return groupName;
+  }
+
+  @Override
+  public String toString() {
+    return "ScatterReceiver:" + Utils.simpleName(groupName) + ":" + Utils.simpleName(operName) + ":" + version;
+  }
+
+  @Override
+  public void onNext(final GroupCommunicationMessage msg) {
+    topology.handle(msg);
+  }
+
+  @Override
+  public List<T> receive() throws NetworkException, InterruptedException {
+    LOG.entering("ScatterReceiver", "receive", this);
+    // I am intermediate node or leaf.
+    LOG.fine("I am " + this);
+
+    if (init.compareAndSet(false, true)) {
+      LOG.fine(this + " Communication group initializing");
+      commGroupClient.initialize();
+      LOG.fine(this + " Communication group initialized");
+    }
+
+    final List<T> retList;
+    try {
+      LOG.fine(this + " Waiting to receive scatter from parent");
+      final byte[] data = topology.recvFromParent(ReefNetworkGroupCommProtos.GroupCommMessage.Type.Scatter);
+
+      LOG.fine(this + " Trying to send scatter to children");
+      topology.sendToChildren(data, ReefNetworkGroupCommProtos.GroupCommMessage.Type.Scatter);
```

Comment:
Sorry, I think I'm the one who misunderstood our conversation on JIRA. I'll change the code so that  we use a encoded topology to send only relevant data.